plugins/scroll-options/src/AutoScroll.js

```diff
@@ -0,0 +1,169 @@
+      const dt = now - this.lastTime_;
+      this.lastTime_ = now;
+      // This method could be called multiple times per ms, and we only want to
+      // scroll if we should actually move.
+      if (dt > 0) {
+        this.scrollTick_(dt);
+      }
+
+      this.animationFrameId_ =
+          requestAnimationFrame(this.nextAnimationStep_.bind(this));
+    }
+  }
+
+  /**
+   * Perform scroll given time passed.
+   * @param {number} msPassed Number of ms since last scroll tick.
+   * @private
+   */
+  scrollTick_(msPassed) {
+    const scrollDx = this.activeScrollVector_.x * msPassed;
+    const scrollDy = this.activeScrollVector_.y * msPassed;
+    this.scrollWorkspaceWithBlock(scrollDx, scrollDy);
+  }
+
+  /**
+   * Scrolls the workspace the given amount during a block drag.
+   * Also updates the dragger based on the amount actually scrolled.
+   * @param {number} scrollDx Amount to scroll in horizontal direction.
+   * @param {number} scrollDy Amount to scroll in vertical direction.
+   */
+  scrollWorkspaceWithBlock(scrollDx, scrollDy) {
+    const oldLocation = this.getDragSurfaceLocation_();
+
+    // As we scroll, we shouldn't expand past the content area that existed
+    // before the block was picked up. Therefore, we use cached ContentMetrics
+    // so that the content area does not change as we scroll.
+    const metricsManager =
+    /** @type {ScrollMetricsManager} */ (
+        this.workspace_.getMetricsManager());
+    metricsManager.useCachedContentMetrics = true;
+    const newX = this.workspace_.scrollX + scrollDx;
+    const newY = this.workspace_.scrollY + scrollDy;
+    this.workspace_.scroll(newX, newY);
+    metricsManager.useCachedContentMetrics = false;
+
+    const newLocation = this.getDragSurfaceLocation_();
+
+    // How much we actually ended up scrolling.
+    const deltaX = newLocation.x - oldLocation.x;
+    const deltaY = newLocation.y - oldLocation.y;
+
+    // If we did scroll, we need to let the dragger know by how much.
+    // The dragger will update its values so that things like connection
+    // markers will stay consistent.
+    if (deltaX || deltaY) {
+      this.workspace_.currentGesture_.getCurrentDragger()
+          .moveBlockWhileDragging(deltaX, deltaY);
+    }
+  }
+
+  /**
+   * Gets the current location of the drag surface.
+   * This has to return a copy to work.
+   * TODO: Deduplicate with index.js.
+   * @return {!Blockly.utils.Coordinate} The current coordinate.
+   * @private
+   */
+  getDragSurfaceLocation_() {
+    const dragSurface = this.workspace_.getBlockDragSurface();
+    const workspaceOffset = dragSurface.getWsTranslation();
+    return new Blockly.utils.Coordinate(workspaceOffset.x, workspaceOffset.y);
+  }
+
+  /**
+   * Updates the scroll vector for the current autoscroll and begins the
+   * animation.
+   * @param {!Blockly.utils.Coordinate} scrollVector New scroll velocity vector
+   *     in pixels per ms.
+   */
+  updateProperties(scrollVector) {
```

Comment:
This could get called in the middle of an animation. For example, you are holding the block at the top-left corner, and while scrolling, you move the block to just the top of the screen. In that case, you're not really starting the animation from scratch, at least from the perspective of the block dragger, where this is called.

In the original version, it was `updateProperties` because it also passed the mouseX and mouseY values. I'm not yet sure whether I'll need them in the next phase of smarter auto-dragging because I haven't tried to implement them yet. If not, I could possibly rename this then, but I'm not sure `startAnimation` is quite accurate.

I think I need to look into not requesting a new animation frame if we're already in the middle of one, but I'll do that separately.

Reply:
Okay, I changed it so it doesn't start a new animation frame unless we haven't had one yet. So now it truly is just updating the properties, usually, as opposed to starting a new one.

It's also way smoother now that it's working properly instead of basically starting a new animation with every drag step. Yay!